docs/custom_promise_types/modules.md

```diff
@@ -0,0 +1,367 @@
+        if not promiser.startswith("/"):
+            raise ValidationError(f"File path '{promiser}' must be absolute")
+        for name, value in attributes.items():
+            if name != "repo":
+                raise ValidationError(f"Unknown attribute '{name}' for git promises")
+            if name == "repo" and type(value) is not str:
+                raise ValidationError(f"'repo' must be string for git promise types")
+
+    def evaluate_promise(self, promiser, attributes):
+        if not promiser.startswith("/"):
+            raise ValidationError("File path must be absolute")
+
+        folder = promiser
+        url = attributes["repo"]
+
+        if os.path.exists(folder):
+            self.promise_kept()
+            return
+
+        self.log_info(f"Cloning '{url}' -> '{folder}'...")
+        os.system(f"git clone {url} {folder} 1>/dev/null 2>/dev/null")
+
+        if os.path.exists(folder):
+            self.log_info(f"Successfully cloned '{url}' -> '{folder}'")
+            self.promise_repaired()
+        else:
+            self.log_error(f"Failed to clone '{url}' -> '{folder}'")
+            self.promise_not_kept()
+
+
+if __name__ == "__main__":
+    GitPromiseTypeModule().start()
+```
+
+### Logging
+
+Log levels in CFEngine are explained here:
+https://github.com/cfengine/core/blob/master/CONTRIBUTING.md#log-levels
+
+In short, when writing a promise module, these log levels should be used:
+
+* `critical` - Serious errors in protocol or module itself (not in policy)
+* `error` - Errors when validating / evaluating a promise, including syntax errors and promise not kept
+* `info` - Changes made (repaired promises)
+* `verbose` - Human understandable detailed information about promise evaluation
+* `debug` - Programmer-level information that is only useful for CFEngine developers or module developers
+
+### Results
+
+Each operation performed by the module, sends a result back to the agent.
+The possible results are as follows:
+
+* Shared between operations:
+  * `error` - an unexpected error occured in the module or protocol, indicating a bug in CFEngine or the promise module
+    * Should be explained by a `critical` level log message
+* Promise validation:
+  * `valid` - No problems with the data or data types in promise
+  * `invalid` - There are problems with the promise data or data types
+    * Should be explained by an `error` level log message
+* Promise evaluation:
+  * `kept` - promise satisfied already, no change made
+  * `repaired` - promise not satisfied before, but fixed now
+    * The change should be explained in a `info` level log message
+  * `not_kept` - promise not satisfied before, and could not be fixed
+    * Should be explained by an `error` level log message
+* Teminate:
+  * `success` - Module succesfully terminated without errors
+  * `failure` - There were problems when trying to clean up / terminate
+    * Should be explained by a `critical` level log message
```

Comment:
Is it important that this is handled at the protocol level? Can it instead be handled as a result of the return code? This allows things like segfaults also top be covered by the same mechanism.

Reply:
The simplicity of using the exit code is tempting, but then it's more difficult to get meaningful log messages in case of errors. This approach gives us more flexibility for modules to give good error messages, and we don't have to "invent" another system for how to log those.

Reply:
Right. It also just occurred to me that using exit codes would conflict with the goal of having long running module processes, so the this approach is definitely better.